### PR TITLE
added Error Dialogs in all UI Components to inform the User if an Error occurs

### LINKED ISSUE #47: User is not informed about any Errors
UI needs to implement a window which every class can access, which is able to inform the User about an Error.

### PR DESCRIPTION
**Pull request by aleksn41**

### CODE REVIEW

File: app/src/main/java/de/dhbw/corona_world_app/ui/statistic/StatisticFragment.java [hunk 1]
```diff
@@ -40,24 +44,18 @@ public View onCreateView(@NonNull LayoutInflater inflater,
...
                 new ViewModelProvider(requireActivity()).get(StatisticCallViewModel.class);
 
-        //TODO remove this
-        if (!statisticCallViewModel.isInit()) {
+        if (statisticCallViewModel.isNotInit()) {
             try {
                 statisticCallViewModel.init(requireActivity().getFilesDir(), ThreadPoolHandler.getInstance());
             } catch (IOException e) {
-                e.printStackTrace();
+                ErrorDialog.createBasicErrorDialog(getContext(), "We were not able to create new Files", "Please restart the application or it will not function properly", null);
```
Comment (Tommy-FF) [inline app/src/main/java/de/dhbw/corona_world_app/ui/statistic/StatisticFragment.java]: Very nice error message but I think a "Please restart the app" without the second part suffices

Comment (Tommy-FF) [reply app/src/main/java/de/dhbw/corona_world_app/ui/statistic/StatisticFragment.java]: We should discuss if we make standardized error codes (e.g. "Error F212" or something along those lines)

File: app/src/main/java/de/dhbw/corona_world_app/ui/statistic/StatisticFragment.java [hunk 2]
```diff
@@ -74,17 +72,16 @@ public void onViewCreated(@NotNull View view, @Nullable Bundle savedInstanceStat
...
-            statisticCallViewModel.addData(Collections.singletonList(request));
+            statisticCallViewModel.addData(Collections.singletonList(request)).get();
             statisticCallViewModel.saveAllData();
-        } catch (ExecutionException e) {
-            e.printStackTrace();
-        } catch (InterruptedException e) {
-            e.printStackTrace();
+        } catch (ExecutionException | InterruptedException e) {
+            Log.e(TAG,"could not save Data",e);
+            Toast.makeText(getContext(), "Could not save Data of this Session, please restart the Application if you want your Session to be saved", Toast.LENGTH_LONG).show();
```
Comment (Tommy-FF) [inline app/src/main/java/de/dhbw/corona_world_app/ui/statistic/StatisticFragment.java]: Same as above...

File: app/src/main/java/de/dhbw/corona_world_app/ui/tools/StatisticCallRecyclerViewFragment.java [hunk 3]
```diff
@@ -67,7 +64,8 @@ public void deleteItems(Set<Integer> ItemIds) {
                     statisticCallViewModel.deleteItems(ItemIds, getDataType()).get();
                 } catch (ExecutionException | InterruptedException e) {
                     Log.e(this.getClass().getName(), "items could not be deleted", e);
-                    //TODO handle
+                    //this should not happen
+                    ErrorDialog.createBasicErrorDialog(getContext(),"An critical Error has occurred","It seems that something unexpected happened, please restart the App and if that does not help reinstall the App",null);
```
Comment (Tommy-FF) [inline app/src/main/java/de/dhbw/corona_world_app/ui/tools/StatisticCallRecyclerViewFragment.java]: Same problem as above. "Please reinstall" seems a tad extreme... I don't know if a user who gets this message will ever reinstall the app.

File: app/src/main/java/de/dhbw/corona_world_app/ui/tools/StatisticCallRecyclerViewFragment.java [hunk 4]
```diff
@@ -102,23 +100,24 @@ public void onChanged(List<Pair<StatisticCall, Boolean>> pairs) {
...
     @Override
     public void onStop() {
-        super.onStop();
+        Log.d(this.getClass().getName()+"|"+getDataType(),"Stopping Fragment");
         try {
             statisticCallViewModel.saveAllData();
         } catch (ExecutionException | InterruptedException e) {
             Log.e(this.getClass().getName(),"could not save Data",e);
-            //TODO handle
+            Toast.makeText(getContext(), "Could not save Data of this Session, please restart the Application if you want your Session to be saved", Toast.LENGTH_LONG).show();
```
Comment (Tommy-FF) [inline app/src/main/java/de/dhbw/corona_world_app/ui/tools/StatisticCallRecyclerViewFragment.java]: Same problem as above. We should discuss.

File: app/src/main/java/de/dhbw/corona_world_app/ui/tools/StatisticCallRecyclerViewFragment.java [hunk 5]
```diff
@@ -128,4 +127,23 @@ public void onStop() {
...
     public abstract void initViewModelData(StatisticCallViewModel statisticCallViewModel);
+
+    protected void tryRepairingData(){
+        ErrorDialog.createBasicErrorDialog(getContext(), "Your Data seems to be corrupt", "We were not able to save new Data, we will try to fix this Problem", (dialog, which) -> {
+            //TODO implement check to see if Data can be recovered
+            boolean canBeRecovered=false;
+            if(canBeRecovered){
+                //recover Data
+            }else{
+                ErrorDialog.createBasicErrorDialog(getContext(), "We were not able to recover your Data", "Your History and your Favourites must be deleted for this app to function properly, we are so sorry", (dialog1, which1) -> {
```
Comment (Tommy-FF) [inline app/src/main/java/de/dhbw/corona_world_app/ui/tools/StatisticCallRecyclerViewFragment.java]: What is this error message?

File: app/src/main/java/de/dhbw/corona_world_app/ui/tools/StatisticCallRecyclerViewFragment.java [hunk 6]
```diff
@@ -128,4 +127,23 @@ public void onStop() {
     public abstract ShowStatisticInterface getShowStatisticInterface();
 
     public abstract void initViewModelData(StatisticCallViewModel statisticCallViewModel);
+
+    protected void tryRepairingData(){
+        ErrorDialog.createBasicErrorDialog(getContext(), "Your Data seems to be corrupt", "We were not able to save new Data, we will try to fix this Problem", (dialog, which) -> {
```
Comment (Tommy-FF) [inline app/src/main/java/de/dhbw/corona_world_app/ui/tools/StatisticCallRecyclerViewFragment.java]: Fine error message, but as stated above, we should discuss if we replace some of it with a standardized error code/message

File: app/src/main/java/de/dhbw/corona_world_app/ui/tools/StatisticCallRecyclerViewFragment.java [hunk 7]
```diff
@@ -128,4 +127,23 @@ public void onStop() {
...
+            boolean canBeRecovered=false;
+            if(canBeRecovered){
+                //recover Data
+            }else{
+                ErrorDialog.createBasicErrorDialog(getContext(), "We were not able to recover your Data", "Your History and your Favourites must be deleted for this app to function properly, we are so sorry", (dialog1, which1) -> {
+                    try {
+                        statisticCallViewModel.deleteAllItems().get();
+                    } catch (ExecutionException | InterruptedException e1) {
+                        Log.e(this.getClass().getName()+"|"+getDataType(),"Not able to delete corrupt Data", e1);
+                        ErrorDialog.createBasicErrorDialog(getContext(),"There has been an error deleting your Data","Something has gone terribly wrong, please reinstall the app and try again",null);
```
Comment (Tommy-FF) [inline app/src/main/java/de/dhbw/corona_world_app/ui/tools/StatisticCallRecyclerViewFragment.java]: Look above. "Reinstall the app" should not be a message we want to give the user...

**Review by Tommy-FF — CHANGES_REQUESTED** [review]
We should discuss how we handle the messaging.

**Comment by aleksn41** [comment]
Error Codes are now standardized and displayed to User in all UI related Errors

**Review by Tommy-FF — APPROVED** [review]

**Pull request merged**